### PR TITLE
Add JDBC template support for DAOs.

### PR DESCRIPTION
**Pull request by dewniMW**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.uma.common/src/main/java/org/wso2/carbon/identity/oauth/uma/common/JdbcUtils.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import javax.sql.DataSource;
+import javax.xml.namespace.QName;
+
+/**
+ * This class checks whether the database is created, if not it creates one. It reads
+ * the data source properties from the identity.xml.
+ */
+public class JdbcUtils {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.uma.common/src/main/java/org/wso2/carbon/identity/oauth/uma/common/JdbcUtils.java]: Don't we have utils in identity framework to find the value for this property. Check the OAuth DAOs to get an idea

Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.common/src/main/java/org/wso2/carbon/identity/oauth/uma/common/JdbcUtils.java]: There was no method to return the data source at JDBCPersistenceManager. Therefore a method was added to return the data source https://github.com/wso2/carbon-identity-framework/pull/1617

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/PermissionTicketDAO.java [hunk 2]
```diff
@@ -68,138 +68,153 @@
...
+                );
+                if (resourceId == null) {
+                    throw new UMAClientException(UMAConstants.ErrorMessages
+                            .ERROR_BAD_REQUEST_INVALID_RESOURCE_ID, "Permission request failed with bad resource ID : "
+                            + resource.getResourceId());
                 }
-            } catch (SQLException e) {
+            } catch (DataAccessException e) {
                 throw new UMAServerException(UMAConstants.ErrorMessages
                         .ERROR_INTERNAL_SERVER_ERROR_FAILED_TO_PERSIST_REQUESTED_PERMISSIONS, e);
```
[the remark this answers is not in the document]
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.permission.service/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/PermissionTicketDAO.java]: will add a meaningful error message

File: components/org.wso2.carbon.identity.oauth.uma.common/src/main/java/org/wso2/carbon/identity/oauth/uma/common/JdbcUtils.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.wso2.carbon.database.utils.jdbc.NamedJdbcTemplate;
+import org.wso2.carbon.identity.core.persistence.JDBCPersistenceManager;
+
+/**
+ * This class checks whether the database is created, if not it creates one. It reads
+ * the data source properties from the identity.xml.
+ */
+public class JdbcUtils {
+
+    public static NamedJdbcTemplate getNewNamedTemplate() {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.common/src/main/java/org/wso2/carbon/identity/oauth/uma/common/JdbcUtils.java]: Add java doc comment

File: components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImpl.java [hunk 4]
```diff
@@ -127,11 +140,11 @@ private boolean isValidTokenScope(MessageContext context) {
     }
 
     /**
-     * Retrieve resource owner name
+     * Retrieve resource owner name with user domain
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImpl.java]: Add a full stop at the end

File: components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImplTest.java [hunk 5]
```diff
@@ -115,25 +122,29 @@ public void testRegisterPermission() throws Exception {
         when(mockHTTPServletRequest.getAttribute(anyString())).thenReturn(mockAuthenticationContext);
         String[] tokenScopes = new String[]{PAT_SCOPE};
         when(mockAuthenticationContext.getParameter(anyString())).thenReturn(tokenScopes);
-        when(mockAuthenticationContext.getUser()).thenReturn(new User());
+        when(mockAuthenticationContext.getUser()).thenReturn(mockUser);
+        when(mockUser.getUserName()).thenReturn("PRIMARY/admin");
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImplTest.java]: Move "PRIMARY/admin" to a constant and reuse in other places

File: components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImplTest.java [hunk 6]
```diff
@@ -115,25 +122,29 @@ public void testRegisterPermission() throws Exception {
...
+        when(UserCoreUtil.removeDomainFromName(anyString())).thenReturn("admin");
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImplTest.java]: Move "admin" to a constant and reuse in other places

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/service/PermissionService.java [hunk 7]
```diff
@@ -30,6 +30,8 @@
  */
 public interface PermissionService {
 
-    PermissionTicketModel issuePermissionTicket(List<Resource> resourceList, int tenantId, String resourceOwnerName)
-            throws UMAClientException, UMAServerException;
+    PermissionTicketModel issuePermissionTicket(List<Resource> resourceList, int tenantId, String resourceOwnerName,
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.service/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/service/PermissionService.java]: Add method comment

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/TestConstants.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.oauth.uma.permission.service;
+
+public class TestConstants {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/TestConstants.java]: Add class comment

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java [hunk 9]
```diff
@@ -47,7 +50,7 @@
     private static final String STORE_PT_RESOURCE_SCOPES_QUERY = "INSERT INTO IDN_PT_RESOURCE_SCOPE " +
             "(ID, PT_RESOURCE_ID, PT_SCOPE_ID) VALUES (?, ?, ?)";
 
-    protected void initiateH2Base(String databaseName, String scriptPath) throws Exception {
+    public static void initiateH2Base(String databaseName, String scriptPath) throws Exception {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java]: Add method comment

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java [hunk 10]
```diff
@@ -60,7 +63,7 @@ protected void initiateH2Base(String databaseName, String scriptPath) throws Exc
         dataSourceMap.put(databaseName, dataSource);
     }
 
-    protected void closeH2Base(String databaseName) throws Exception {
+    public static void closeH2Base(String databaseName) throws Exception {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java]: Add method comment

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java [hunk 11]
```diff
@@ -76,6 +79,13 @@ public static Connection getConnection(String database) throws SQLException {
         throw new RuntimeException("No datasource initiated for database: " + database);
     }
 
+    public static Connection spyConnection(Connection connection) throws SQLException {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java]: Add method comment

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java [hunk 12]
```diff
@@ -93,8 +103,8 @@ public static BasicDataSource getDatasource(String datasourceName) {
         throw new RuntimeException("No datasource initiated for database: " + datasourceName);
     }
 
-    protected void createPTTable(String databaseName, long id, String pt, Timestamp timecreated, long period,
-                                 String state, int tenantId) throws Exception {
+    public static void createPTTable(String databaseName, long id, String pt, Timestamp timecreated, long period,
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java]: Add method comment

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java [hunk 13]
```diff
@@ -113,7 +123,8 @@ protected void createPTTable(String databaseName, long id, String pt, Timestamp
         }
     }
 
-    protected void createPTResourceTable(String databaseName, long id, long ptResourceId, long ptId) throws Exception {
+    public static void createPTResourceTable(String databaseName, long id, long ptResourceId, long ptId) throws
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java]: Add method comment

File: components/org.wso2.carbon.identity.oauth.uma.resource.endpoint/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/endpoint/impl/ResourceRegistrationApiServiceImpl.java [hunk 14]
```diff
@@ -193,32 +172,24 @@ public Response getResourceIds(MessageContext context) {
...
-                    log.debug("Access token is valid and all required context information " +
-                            "are existing in the token.");
-                }
-                if (isResourceIdValid(resourceId)) {
-
+            if (isValidTokenScope(context) && isResourceIdExists(resourceId)) {
+                if (updatedResource.getResource_Scopes().isEmpty()) {
                     if (log.isDebugEnabled()) {
-                        log.debug("Valid resourceId :" + resourceId);
+                        log.debug("Cannot update resource: " + resourceId + "as resource scopes are not provided");
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.resource.endpoint/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/endpoint/impl/ResourceRegistrationApiServiceImpl.java]: Change to `log.debug("Cannot update resource: " + resourceId + " as resource scopes are not provided.");`

File: components/org.wso2.carbon.identity.oauth.uma.resource.service/src/test/java/org/wso2/carbon/identity/oauth/uma/resource/service/impl/ResourceServiceImplTest.java [hunk 15]
```diff
@@ -57,14 +54,16 @@ public IObjectFactory getObjectFactory() {
     public void testRegisterResource() throws Exception {
 
         mockStatic(ResourceDAO.class);
-        assertNull(resourceService.registerResource(resource, resourceOwnerName, tenantId, consumerKey));
+        assertNull(resourceService.registerResource(resource, "admin", -1234,
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.resource.service/src/test/java/org/wso2/carbon/identity/oauth/uma/resource/service/impl/ResourceServiceImplTest.java]: Change variables to constants and reuse in other places as well.

File: components/org.wso2.carbon.identity.oauth.uma.resource.service/src/test/java/org/wso2/carbon/identity/oauth/uma/resource/service/dao/util/DAOUtils.java [hunk 16]
```diff
@@ -27,12 +27,15 @@
 import java.util.HashMap;
 import java.util.Map;
 
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.spy;
+
 public class DAOUtils {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.resource.service/src/test/java/org/wso2/carbon/identity/oauth/uma/resource/service/dao/util/DAOUtils.java]: Add class comment and method comments.

File: components/org.wso2.carbon.identity.oauth.uma.resource.service/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/service/dao/ResourceDAO.java [hunk 17]
```diff
@@ -54,182 +56,138 @@
...
+            namedJdbcTemplate.executeQuery(SQLQueries.GET_RESOURCE_SCOPES, (resultSet, rowNumber) ->
+                    scopes.add(resultSet.getString(1)), namedPreparedStatement ->
+                    namedPreparedStatement.setInt(UMAConstants.SQLPlaceholders.ID, id));
+
+            resource.setScopes(scopes);
+
+            namedJdbcTemplate.executeQuery(SQLQueries.GET_RESOURCE_META_DATA, (resultSet, rowNumber) -> {
+                        if (resultSet.getString("PROPERTY_KEY").equals(ICON_URI)) {
+                            resource.setIconUri(resultSet.getString(
+                                    "PROPERTY_VALUE"));
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.resource.service/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/service/dao/ResourceDAO.java]: Use constants for "PROPERTY_KEY" and "PROPERTY_VALUE". Also, reuse those where needed.

File: components/org.wso2.carbon.identity.oauth.uma.resource.service/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/service/ResourceService.java [hunk 18]
```diff
@@ -30,15 +30,15 @@
 
     boolean deleteResource(String resourceId) throws UMAServerException, UMAClientException;
 
-    List<String> getResourceIds(String resourceOwnerName, String consumerKey) throws UMAServerException
-            , UMAClientException;
+    List<String> getResourceIds(String resourceOwnerName, String userDomain, String consumerKey) throws
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.uma.resource.service/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/service/ResourceService.java]: Add method comments to all methods.

Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.resource.service/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/service/ResourceService.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.resource.service/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/service/dao/ResourceDAO.java (hunk 17 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.resource.service/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/service/dao/ResourceDAO.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.resource.service/src/test/java/org/wso2/carbon/identity/oauth/uma/resource/service/dao/util/DAOUtils.java (hunk 16 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.resource.service/src/test/java/org/wso2/carbon/identity/oauth/uma/resource/service/dao/util/DAOUtils.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.resource.service/src/test/java/org/wso2/carbon/identity/oauth/uma/resource/service/impl/ResourceServiceImplTest.java (hunk 15 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.resource.service/src/test/java/org/wso2/carbon/identity/oauth/uma/resource/service/impl/ResourceServiceImplTest.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.resource.endpoint/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/endpoint/impl/ResourceRegistrationApiServiceImpl.java (hunk 14 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.resource.endpoint/src/main/java/org/wso2/carbon/identity/oauth/uma/resource/endpoint/impl/ResourceRegistrationApiServiceImpl.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java (hunk 9 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/dao/utils/DAOTestUtils.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/TestConstants.java (hunk 8 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.permission.service/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/service/TestConstants.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.permission.service/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/service/PermissionService.java (hunk 7 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.permission.service/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/service/PermissionService.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImplTest.java (hunk 6 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImplTest.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImplTest.java (hunk 5 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/test/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImplTest.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImpl.java (hunk 4 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.permission.endpoint/src/main/java/org/wso2/carbon/identity/oauth/uma/permission/endpoint/PermissionApiServiceImpl.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

File: components/org.wso2.carbon.identity.oauth.uma.common/src/main/java/org/wso2/carbon/identity/oauth/uma/common/JdbcUtils.java (hunk 3 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.uma.common/src/main/java/org/wso2/carbon/identity/oauth/uma/common/JdbcUtils.java]: fixed: https://github.com/wso2-extensions/identity-oauth-uma/pull/68

**Pull request merged**